Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -2246,6 +2247,36 @@ void FrameLoader::commitProvisionalLoad()
         frame->document() ? frame->document()->url().stringCenterEllipsizedToLength().utf8().data() : "",
         pdl ? pdl->url().stringCenterEllipsizedToLength().utf8().data() : "<no provisional DocumentLoader>", cachedPage.get());
 
+    if (RefPtr document = m_frame->document()) {
+        bool canTriggerCrossDocumentViewTransition = false;
+        RefPtr<NavigationActivation> activation;
+        if (pdl) {
+            canTriggerCrossDocumentViewTransition = pdl->navigationCanTriggerCrossDocumentViewTransition(*document);
+
+            RefPtr domWindow = document->domWindow();
+            auto navigationAPIType = pdl->triggeringAction().navigationAPIType();
+            if (domWindow && navigationAPIType) {
+                // FIXME: The NavigationActivation for pageswap should be created after the global
+                // history update, but before the unload event (which might be delayed). Those steps
+                // are currently intertwined, so this creates a fake/detached new history entry to
+                // use for this purpose.
```

Comment:
Is it difficult to separate the unload event from `addIfCacheable()`? Maybe this is okay for now, but it seems very weird to have this situation where we need to create a detached history item with the provisional loader.

Reply:
The answer is that I really don't know at all :) 

Some loads result in the unload event happening deep inside `addIfCacheable`, others happen when we call `transitionToCommitted` into `closeURL()` (which is what the comments say is the expected time).

This code has all sorts of interesting early exit paths where calling a JS event results in the navigation state being changed recursively, and it's really hard to reason about all the possible fallout, especially if there are nested <iframes>.

I couldn't agree more that it's weird, and would be good to unwind this code to match the spec description better, but that seems like a large and danger-fraught task :)